### PR TITLE
[uss_qualifier] Validate that some queries to service_providers are properly authenticated for NET0210

### PR DESCRIPTION
**Pull request by Shastick**
Add check for NET0210, which is about checking that service providers correctly authenticate requests.

This is done via a new `misbehavior` scenario that will inject flight data and query the SP's endpoint for that data without providing credentials.

It behaves very similarly to the `nominal_behavior` scenario, with these two tweaks:
* a DSS instance is required instead of being optional
* the polling stops as soon as we have validated that the SP properly authenticates the flights and flight_details requests.

### CODE REVIEW

File: monitoring/monitorlib/fetch/rid.py [hunk 1]
```diff
@@ -795,6 +819,19 @@ def uss_flights(
...
+                flights_url,
+                params=params,
+            )
+            if failed_q.response.code == 200:
+                verify_sp_checks_credentials.record_failed(
+                    "unauthenticated request was fulfilled",
+                    severity=Severity.MEDIUM,
+                    details=f"queried {flights_url} with no credentials, expected a failure but got a success reply",
+                )
         return FetchedUSSFlights(v22a_query=query)
```
Comment (Shastick) [inline monitoring/monitorlib/fetch/rid.py:835]: I guess that an alternative to doing the unauthenticated query in here would be to do it from `all_flights()` is called based on the `FetchedUSSFlights` that is returned?

File: monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+### flights_data
+
+A [`FlightDataResource`](../../../../resources/netrid/flight_data.py) containing 1 nominal flight per SP under test.
+
+### service_providers
+
+A set of [`NetRIDServiceProviders`](../../../../resources/netrid/service_providers.py) to be tested for proper request authentication. This scenario requires at least one SP under test.
+
+### observers
```
Comment (BenjaminPelletier) [inline monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md:17]: Why is this needed?

File: monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+Per **[interuss.automated_testing.rid.injection.UpsertTestSuccess](../../../../requirements/interuss/automated_testing/rid/injection.md)**, the injection attempt of the valid flight should succeed for every NetRID Service Provider under test.
+
+**[astm.f3411.v19.NET0500](../../../../requirements/astm/f3411/v19.md)** requires a Service Provider to provide a persistently supported test instance of their implementation.
+This check will fail if the flight was not successfully injected.
+
+#### Identifiable flights check
+
+This particular test requires each flight to be uniquely identifiable by its 2D telemetry position; the same (lat, lng) pair may not appear in two different telemetry points, even if the two points are in different injected flights.  This should generally be achieved by injecting appropriate data.
+
+### Unauthenticated requests test step
```
Comment (BenjaminPelletier) [inline monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md:46]: Let's describe what this test step does prior to the check (attempts to repeat the previously-successful queries, but without authentication this time)

File: monitoring/uss_qualifier/scenarios/astm/netrid/v19/nominal_behavior.md [hunk 4]
```diff
@@ -162,6 +162,10 @@ Taking into account the propagation time of the injected flights, if the total n
 For a display area with a diagonal greather than *NetDetailsMaxDisplayAreaDiagonal* and less than *NetMaxDisplayAreaDiagonal*, **[astm.f3411.v19.NET0480](../../../../requirements/astm/f3411/v19.md)** requires that a Display provider shall cluster UAs in close proximity to each other using a circular or polygonal area covering no less than *NetMinClusterSize* percent of the display area size.
 This check validates that the display area of a cluster, measured and provided in square meters by the test harness, is no less than *NetMinClusterSize* percent of the display area.
 
+#### Missing credentials check
```
Comment (BenjaminPelletier) [inline monitoring/uss_qualifier/scenarios/astm/netrid/v19/nominal_behavior.md:165]: Why is this here?

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        self._evaluation_configuration = evaluation_configuration
+        self._injected_flights = []
+        self._injected_tests = []
+        self._dss_pool = dss_pool
+
+    @property
+    def _rid_version(self) -> RIDVersion:
+        return (
+            self._dss_pool.dss_instances[0].rid_version
+            if self._dss_pool
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py:85]: This file is essentially a copy-paste of `nominal_behavior.py`: should I also fix things there?

In this particular case, what would be the proper way to determine the RIDVersion?

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            relevant_past_data_period=self._rid_version.realtime_period
+            + config.max_propagation_latency.timedelta,
+        )
+
+    def _poll_unauthenticated_during_flights(self):
+        config = self._evaluation_configuration.configuration
+
+        # Evaluate observed RID system states
+        evaluator = display_data_evaluator.RIDObservationEvaluator(
+            self,
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py:203]: Inherited from copy-paste: will also fix in `nominal_behavior.py`

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+import uuid
+import time
+from typing import List, Optional
+
+import arrow
+from implicitdict import ImplicitDict
+from loguru import logger
+from requests.exceptions import RequestException
+
+from monitoring.monitorlib import fetch
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py:12]: Did a cleanup pass. Somehow thought the linter handles it but it looks like I was mistaken :)

File: monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py [hunk 8]
```diff
@@ -216,6 +218,78 @@ def __init__(
                 f"Cannot evaluate a system using RID version {rid_version} with a DSS using RID version {dss.rid_version}"
             )
 
+    def evaluate_and_test_authentication(
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:221]: No particular reason: it also started from a copy-paste, I'll move it to the misbehavior file

File: monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py [hunk 9]
```diff
@@ -194,7 +196,7 @@ def __init__(
         injected_flights: List[InjectedFlight],
         config: EvaluationConfiguration,
         rid_version: RIDVersion,
-        dss: Optional[DSSInstance] = None,
+        dss: DSSInstance,
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:199]: Ah ok: I can make the `dss` optional as it is in the `nominal_behavior` case, but this would effectively make this scenario a no-op if no DSS is defined.


Would that be fine? 

I was expecting that users of the qualifier would simply disable tests that made no sense for their use-case.

File: monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py [hunk 10]
```diff
@@ -216,6 +218,78 @@ def __init__(
...
+            self._test_scenario.record_note(
+                "unauthenticated_queries",
+                f"about to repeat {len(queries_to_repeat)} flights queries without credentials",
+            )
+
+            # Attempt to re-query the flights and flight details URLs:
+            for fq in list(sp_observation.uss_flight_queries.values()) + list(
+                sp_observation.uss_flight_details_queries.values()
+            ):
+                failed_q = fetch.query_and_describe(
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:270]: As far as I can tell, GET params are preserved in `fq.query.request.url`.

For situations where things are passed in the body with a POST I'll pass the `json` and `body` fields as well.

File: monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md (hunk 2 above)
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/v19/misbehavior.md:17]: I started with the assumption that I'd need the observer to access the SP, which turned out to be false -> will remove

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py]: Upon closer inspection: the type hint seems to be wrong, and `test_scenario` should be a `TestScenario`: passing `type(self)` causes an exception [here](https://github.com/interuss/monitoring/blob/0800a9b9e14b0b298170b1b12390b6298b080c2f/monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py#L225) as illustrated by [this test run](https://github.com/interuss/monitoring/actions/runs/6157168366/job/16707361920#step:4:2764)

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenjaminPelletier) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py]: Ah, sorry you get to clean up -- yes, someone should update nominal_behavior also, though that could be in a separate PR.  This common implementation gets subclassed in two places: one for F3411-19 and one for F3411-22a.  Each of those subclasses definitely knows which RIDVersion we should be using.  There are multiple ways that information could be transmitted, but I think my personally-preferred way would probably be something like:

```python
class Misbehavior(GenericTestScenario):
    ...
    def _rid_version(self) -> RIDVersion:
        raise NotImplementedError("Misbehavior test scenario subclass must specify _rid_version")

<in v19:>
class Misbehavior(TestScenario, CommonMisbehavior):
    def _rid_version(self) -> RIDVersion:
        return RIDVersion.f3411_19

<in v22a:>
class Misbehavior(TestScenario, CommonMisbehavior):
    def _rid_version(self) -> RIDVersion:
        return RIDVersion.f3411_22a
```

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BenjaminPelletier) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py]: Sorry, yes, I was misled by IntelliJ's indication -- TestScenarioType is supposed to indicate "A concrete instance of a class that inherits from TestScenario".  In any case, the current content is fine and my comment was incorrect.

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (BenjaminPelletier) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/misbehavior.py:13]: Yeah, it would be nice if the formatter handled it, but it's a little tricky in Python since sometimes it's useful and important to import something without using it (we do that in mock_uss to get the Flask decorators to attach to the handler routines)

File: monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (BenjaminPelletier) [reply monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py]: Ah, got it -- sounds good.

**Comment by Shastick** [comment]
Unsure if this is flaky or due to interfering tests, but I ran into multiple of these failures [on the CI](https://github.com/interuss/monitoring/actions/runs/6164370000/job/16729944465#step:4:1998):

```
2023-09-12 20:15:38.239 | WARNING  | monitoring.uss_qualifier.suites.suite:_print_failed_check:50 - New failed check:
  details: 'Flight be73f2ee-528f-4d8d-8725-eb0de98d544a: has no telemetry position that
    belong to a cluster.
  
    Severity Medium upgraded to Critical because USS_QUALIFIER_STOP_FAST environment
    variable indicates true'
  documentation_url: https://github.com/interuss/monitoring/blob/889863e731c55d197ddb13c9d2d7faf3779a3cf0/monitoring/uss_qualifier/scenarios/astm/netrid/v19/nominal_behavior.md#individual-flights-obfuscation-check
  name: Individual flights obfuscation
  participants:
  - uss3
  requirements:
  - astm.f3411.v19.NET0490
  severity: Critical
  summary: Error while evaluating obfuscation of individual flights. Flight was outside
    of all clusters.
  timestamp: '2023-09-12T20:15:38.237596Z'
```

Will have a look tomorrow.

**Comment by BenjaminPelletier** [comment]
> Unsure if this is flaky or due to interfering tests, but I ran into multiple of these failures [on the CI](https://github.com/interuss/monitoring/actions/runs/6164370000/job/16729944465#step:4:1998):
> ...
> Will have a look tomorrow.

I think this is likely to be #144 and if so should be transient -- I'd poke @mickmis next week.  It looks like the CI is passing now.

**Review by BenjaminPelletier — APPROVED** [review]
